<h1>Lab | Inferential statistics - T-test & P-value</h1>

<h2>Instructions</h2>

1) **One tailed t-test** - In a packing plant, a machine packs cartons with jars. It is supposed that a new machine will pack faster on the average than the machine currently used. To test that hypothesis, the times it takes each machine to pack ten cartons are recorded. The results, in seconds, are shown in the tables in the file files_for_lab/machine.txt. Assume that there is sufficient evidence to conduct the t test, does the data provide sufficient evidence to show if one machine is better than the other?

2) **Matched Pairs Test** - In this challenge we will compare dependent samples of data describing our Pokemon (file files_for_lab/pokemon.csv). Our goal is to see whether there is a significant difference between each Pokemon's defense and attack scores. Our hypothesis is that the defense and attack scores are equal. Compare the two columns to see if there is a statistically significant difference between them and comment your result.

In [1]:
import pandas as pd
import numpy as np
import scipy.stats as st
import matplotlib.pyplot as plt
import seaborn as sns

In [17]:
data = pd.read_csv('files_for_lab/machine.txt', encoding='UTF-16', sep='\t')
data

New machine      Old machine
0         42.1             42.7
1         41.0             43.6
2         41.3             43.8
3         41.8             43.3
4         42.4             42.5
5         42.8             43.5
6         43.2             43.1
7         42.3             41.7
8         41.8             44.0
9         42.7             44.1

In [32]:
data.shape

(10, 2)

<h2>One tailed t-test</h2>

Let's say that **H0: New machine speed >= Old machine speed**, than

**H1: New machine speed < Old machine speed**

In [23]:
data.columns = ['new_machine', 'old_machine']

Experiments are independent, so:

In [33]:
st.ttest_ind(data['new_machine'], data['old_machine'], equal_var=False)


Ttest_indResult(statistic=-3.397230706117603, pvalue=0.0032422494663179747)

In [34]:
#negative statistic means \mu_a-\mu_b < 0   ---->  \mu_a < \mu_b 

Our statistic is: - 3.397

The p_value corresponding to our statistic is: 0.00324

The significance level is: 0.05

So we reject the H0, therefore New machine speed < Old machine speed.


<h2>Matched Pairs Test</h2>

In [30]:
data2 = pd.read_csv('files_for_lab/pokemon.csv')
data2

#                   Name   Type 1  Type 2  Total  HP  Attack  Defense  \
0      1              Bulbasaur    Grass  Poison    318  45      49       49   
1      2                Ivysaur    Grass  Poison    405  60      62       63   
2      3               Venusaur    Grass  Poison    525  80      82       83   
3      3  VenusaurMega Venusaur    Grass  Poison    625  80     100      123   
4      4             Charmander     Fire     NaN    309  39      52       43   
..   ...                    ...      ...     ...    ...  ..     ...      ...   
795  719                Diancie     Rock   Fairy    600  50     100      150   
796  719    DiancieMega Diancie     Rock   Fairy    700  50     160      110   
797  720    HoopaHoopa Confined  Psychic   Ghost    600  80     110       60   
798  720     HoopaHoopa Unbound  Psychic    Dark    680  80     160       60   
799  721              Volcanion     Fire   Water    600  80     110      120   

     Sp. Atk  Sp. Def  Speed  Generation  Legendary  
0         65       65     45           1      False  
1         80       80     60           1      False  
2        100      100     80           1      False  
3        122      120     80           1      False  
4         60       50     65           1      False  
..       ...      ...    ...         ...        ...  
795      100      150     50           6       True  
796      160      110    110           6       True  
797      150      130     70           6       True  
798      170      130     80           6       True  
799      130       90     70           6       True  

[800 rows x 13 columns]

Let's define that **H0: defense score = attack score**,

than **H1: defense score != attack score** (so one of them is lower/higher than the other)

In [35]:
st.ttest_rel(data2['Attack'], data2['Defense'])

Ttest_relResult(statistic=4.325566393330478, pvalue=1.7140303479358558e-05)

Pvalue (1.7140303479358558e-05) <<< 0.05, so we reject the null hypothesis.

Therefore defense score != attack score (so one of them is lower/higher than the other)

<h2>OPTIONAL PART | Inferential statistics - ANOVA</h2>

Note: The following lab is divided in 2 sections which represent activities 3 and 4.

<h3>Part 1</h3>

In this activity, we will look at another example. Your task is to understand the problem and write down all the steps to set up ANOVA. After the next lesson, we will ask you to solve this problem using Python. Here are the steps that you would need to work on: - Null hypothesis - Alternate hypothesis - Level of significance - Test statistic - P-value - F table

<h3>Context</h3>

Suppose you are working as an analyst in a microprocessor chip manufacturing plant. You have been given the task of analyzing a plasma etching process with respect to changing Power (in Watts) of the plasma beam. Data was collected and provided to you to conduct statistical analysis and check if changing the power of the plasma beam has any effect on the etching rate by the machine. You will conduct ANOVA and check if there is any difference in the mean etching rate for different levels of power. You can find the data anova_lab_data.xlsx file in the files_for_lab folder

- State the null hypothesis
- State the alternate hypothesis
- What is the significance level
- What are the degrees of freedom of model, error terms, and total DoF

Data was collected randomly and provided to you in the table as shown: link to the image - Data: https://education-team-2020.s3-eu-west-1.amazonaws.com/data-analytics/7.05/7.05-lab_data.png 

In [ ]:
# ANOVA test
# to compare the means of multiple groups at the same time

We need to check if there is any difference in the mean etching rate for different levels of power.

**Let's define HO: mean1 = mean2 = mean3 =...meanN**

**Than H1: mean1 != mean2 != mean3 =...meanN**

In [36]:
data3 = pd.read_excel('files_for_lab/anova_lab_data.xlsx')
data3

Power   Etching Rate
0   160 W          5.43
1   180 W          6.24
2   200 W          8.79
3   160 W          5.71
4   180 W          6.71
5   200 W          9.20
6   160 W          6.22
7   180 W          5.98
8   200 W          7.90
9   160 W          6.01
10  180 W          5.66
11  200 W          8.15
12  160 W          5.59
13  180 W          6.60
14  200 W          7.55

In [40]:
data3.columns

Index(['Power ', 'Etching Rate'], dtype='object')

In [42]:
data3['power_count'] = data3.groupby('Power ').cumcount() ##is the new index 
#interest_r
data3_pivot = data3.pivot(index='power_count', columns='Power ', values='Etching Rate')
#interest_r_pivot
data3_pivot.columns = ['Power '+str(x) for x in data3_pivot.columns.values]
data3_pivot.head(10)

Power 160 W  Power 180 W  Power 200 W
power_count                                       
0                   5.43         6.24         8.79
1                   5.71         6.71         9.20
2                   6.22         5.98         7.90
3                   6.01         5.66         8.15
4                   5.59         6.60         7.55

In [47]:
data3_pivot.columns = ['Power_160_W', 'Power_180_W', 'Power_200_W']
data3_pivot.columns

Index(['Power_160_W', 'Power_180_W', 'Power_200_W'], dtype='object')

In [48]:
data3_pivot.mean()

Power_160_W    5.792
Power_180_W    6.238
Power_200_W    8.318
dtype: float64

In [50]:
st.f_oneway(data3_pivot.Power_160_W, data3_pivot.Power_180_W, data3_pivot.Power_200_W)

F_onewayResult(statistic=36.87895470100505, pvalue=7.506584272358903e-06)

Pvalue (7.506584272358903e-06) <<< 0.05, so we reject the null hypothesis.

Therefore there is a difference in the mean etching rate for different levels of power.